frontend/src/host_orchestrator/orchestrator/userartifacts_v1.go

```diff
@@ -0,0 +1,203 @@
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orchestrator
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/google/android-cuttlefish/frontend/src/liboperator/operator"
+)
+
+type UserArtifactChunkV1 struct {
+	// Name of the file
+	Name string
+	// Contains chunk data
+	File io.Reader
+	// SHA1 hash of the whole file data
+	Hash string
+	// Byte offset of the chunk
+	Offset int64
+	// Size of the chunk
+	Size int64
+	// Size of the whole file
+	FileSize int64
+}
+
+// Abstraction for managing user artifacts for launching CVDs. This is for replacing
+// UserArtifactManager, the legacy manager for the same role.
+type UserArtifactsManagerV1 interface {
+	UserArtifactsDirResolver
+	// Update artifact with the passed chunk
+	UpdateArtifact(chunk UserArtifactChunkV1) error
+}
+
+type UserArtifactsManagerV1Opts struct {
+	// Root Directory for storing user artifacts. After Host Orchestrator moves user artifacts into
+	// RootDir, it becomes immutable unless introducing any replacement algorithm(e.g. LRU) on
+	// RootDir to manage the storage.
+	RootDir string
+	// Working directory for storing user artifacts for a while. After all tasks that updating all
+	// chunks are completed, the artifact stored under RootWorkDir is moved into RootDir.
+	RootWorkDir string
+}
+
+// An implementation of the UserArtifactsManagerV1 interface.
+type UserArtifactsManagerV1Impl struct {
+	UserArtifactsManagerV1Opts
+	writeChunkWaitGroups sync.Map
+	moveArtifactMutexes  sync.Map
+	chunkStates          sync.Map
+}
+
+// Creates a new instance of UserArtifactsManagerV1Impl.
+func NewUserArtifactsManagerV1Impl(opts UserArtifactsManagerV1Opts) *UserArtifactsManagerV1Impl {
+	return &UserArtifactsManagerV1Impl{
+		UserArtifactsManagerV1Opts: opts,
+	}
+}
+
+func (m *UserArtifactsManagerV1Impl) GetDirPath(name string) string {
+	return filepath.Join(m.RootDir, name)
+}
+
+func (m *UserArtifactsManagerV1Impl) UpdateArtifact(chunk UserArtifactChunkV1) error {
+	if chunk.Offset < 0 {
+		return operator.NewBadRequestError(fmt.Sprintf("invalid value (chunk_offset:%d)", chunk.Offset), nil)
+	}
+	if chunk.Size < 0 {
+		return fmt.Errorf("invalid value (chunk_size:%d)", chunk.Size)
+	}
+	if chunk.FileSize < 0 {
+		return operator.NewBadRequestError(fmt.Sprintf("invalid value (file_size:%d)", chunk.FileSize), nil)
+	}
+	if chunk.Offset+chunk.Size > chunk.FileSize {
+		return operator.NewBadRequestError(fmt.Sprintf("invalid value pair (chunk_offset:%d, chunk_size:%d, file_size:%d)", chunk.Offset, chunk.Size, chunk.FileSize), nil)
+	}
+	if err := m.writeChunk(chunk); err != nil {
```

Comment:
We are locking on write and later locking on checking hash, I see this impacting upload times. Have you tested uploading a 4GB with 2MB chunks file against a remote machine in GCP using this new API? I recommend doing it to better understand the difference comparing against  a lockless API as the existing one which is our benchmark.

Having said that, I'd like to suggest a less lock-intrusive idea where the only state we keep track of is the bytes written, once the bytes written matches the file size, the server is responsible to verify the checksum and move it to the artifacts directory.